proposals/lip-Introduce_PoA_module.md

```diff
@@ -0,0 +1,607 @@
+* The store key is set to empty bytes.
+* The store value is set to the serialization of an object following `chainPropSchema` below.
+* Notation: Let `chainProperties` be the entry in the chain properties substore.
+
+##### JSON Schema
+
+```java
+chainPropSchema = {
+   "type": "object",
+   "properties": {
+      "roundEndHeight": {
+          "dataType": "uint32",
+          "fieldNumber": 1
+      },
+       "validatorsUpdateNonce": {
+           "dataType": "uint32",
+           "fieldNumber": 2
+      },
+    }
+    "required": [
+       "roundEndHeight",
+       "validatorsUpdateNonce"
+    ]
+}
+```
+
+##### Properties and Default Values
+
+The properties of this schema are as follows:
+
+*   `roundEndHeight`: An integer stating the last height of the round. 
+Its initial value is set after the execution of the genesis block.
+*   `validatorsUpdateNonce`: An integer representing the number of times that the validator set has been updated with an update auhtority command. 
+It is initialized to 0.
+
+### Commands
+
+#### Authority Registration Command
+
+This command is equivalent to the [delegate registration command][delegatedoc] in the DPoS module and has the same schema and similar validity rules. 
+The command ID of this transaction is `COMMAND_ID_REGISTRATION_AUTHORITY`.
+
+##### Parameters
+
+```java
+registrationTransactionParamsSchema = {
+  "type": "object",
+  "properties": {
+    "name": {
+      "dataType": "string",
+      "fieldNumber": 1
+    },
+    "blsKey": {
+      "dataType": "bytes",
+      "fieldNumber": 2
+    },
+    "proofOfPossession": {
+      "dataType": "bytes",
+      "fieldNumber": 3
+    },
+    "generatorKey": {
+      "dataType": "bytes",
+      "fieldNumber": 4
+    }
+  },
+  "required": ["name", "blsKey", "proofOfPossession", "generatorKey"]
+}
+```
+
+##### Verification 
+ 
+Let `trs` be a transaction with module ID `MODULE_ID_POA` and command ID `COMMAND_ID_REGISTRATION_AUTHORITY` to be verified. 
+The list of verification conditions for `trs.params` is as follows:
+
+* The `trs.params.name` property has to contain only characters from the set `[a-z0-9!@$&_.]`, must not be empty and has to be at most `MAX_LENGTH_NAME` characters long.
+* Let `address` be the 20-byte address derived from `trs.senderPublicKey`. Then `address` must not already be registered as a validator. This is, the validator substore has no entry with store key `address`.
+* The value of `trs.params.name` must not already be registered as a validator name. This is, the name substore has no entry with store key `trs.params.name`.
+* `isKeyRegistered(trs.params.blsKey)` must return `true`, where the function is defined in the [Validators module][validatorsModule].
+* `PopVerify(trs.params.blsKey, trs.params.proofOfPossession)` must return `VALID`, where [`PopVerify`][bls-specs-v4-popverify] is part of the [BLS signature scheme][lip-bls-public-key-registration].
+* `tx.params.generatorKey` must have length 32.
```

Comment:
Do we not have similar size constraints for the `blsKey` and `proofOfPossession`?

Reply:
indeed we do, good catch, added.

Reply:
@ManuGowda @ishantiw How is it for other commands, do we always have size checks for all paramters that are enforced via the schema?

So here we are sayting that PopVerify would also include the size checks and the call to this function is also a (stateless) validation. Note that in `registerValidatorKeys` the PoP is verified again so I am not sure that is as intended.

Reply:
I think the LIP has not been updated regarding this comment. In particular, when processing an authority registration command we would verify the PoP twice:
- once during the verification here
- again in the registerValidatorKeys API call

Reply:
I updated the validation of the command to just check that `registerValidatorKeys` does not fail. 
Also added the size checks for every input.

Reply:
I think `registerValidatorKeys` cannot be called in the transaction verification because it induces a state transition. Also, it would now be called twice as the transaction verification also happens when executing the payload.

Reply:
removed the call of `registerValidatorKeys` from the verification step. Now it is only called in the execution step.

Reply:
We still need to check the return value of that function. If it returns `False` the execution must fail.

Reply:
added

Reply:
Looks good. Can be resolved.